src/Steps/Release/MergeUpRelease.php

```diff
@@ -0,0 +1,181 @@
+ * up to 4 - and then optionally delete the 4.12-release branch.
+ */
+class MergeUpRelease extends ReleaseStep
+{
+    public function getStepName()
+    {
+        return 'mergeup';
+    }
+
+    public function run(InputInterface $input, OutputInterface $output)
+    {
+        $this->log($output, "Merging up all modules");
+
+        $this->mergeRecursive($output, $this->getReleasePlan(), $input->getOption('delete-release-branch') ?? false);
+
+        $this->log($output, "All modules merged up");
+    }
+
+    /**
+     * Merge a library and its children up to their major branches
+     *
+     * @param OutputInterface $output
+     * @param LibraryRelease $releasePlanNode Node in release plan being released
+     * @param bool $deleteBranch Whether to delete the release branch after merging
+     */
+    protected function mergeRecursive(OutputInterface $output, LibraryRelease $releasePlanNode, bool $deleteBranch)
+    {
+        $currentBranch = $releasePlanNode->getLibrary()->getBranch();
+
+        // Skip upgrade-only modules and anything not on an actual branch (e.g. using a pre-existing tag)
+        if ($releasePlanNode->getLibrary()->isUpgradeOnly() || !$currentBranch) {
+            return;
+        }
+        $currentLibrary = $releasePlanNode->getLibrary()->getName();
+
+        // Skip if it's not a proper release branch.
+        $currentBranchIsRelease = preg_match(
+            '/' . preg_quote(RewriteReleaseBranches::RELEASE_SUFFIX) . '$/',
+            $currentBranch
+        );
+        if (!$currentBranchIsRelease) {
+            $this->log(
+                $output,
+                "Warning: Current branch '$currentBranch' is not a release branch. "
+                    . "Skipping $currentLibrary and its children.",
+                'bg=yellow'
+            );
+            return;
+        }
+
+        // Do the merge up recursively
+        foreach ($releasePlanNode->getItems() as $item) {
+            $this->mergeRecursive($output, $item, $deleteBranch);
+        }
+        $this->mergeLibrary($output, $releasePlanNode);
+
+        // Optionally delete the branch from the remote
+        if ($deleteBranch) {
+            $this->log(
+                $output,
+                "Deleting branch <info>$currentBranch</info> from remote for <info>{$currentLibrary}</info>"
+            );
+            $releasePlanNode->getLibrary()->getRepository()->run('push', ['--delete', $currentBranch]);
+        }
+    }
+
+    /**
+     * Performs a merge up for a single library
+     *
+     * @param OutputInterface $output
+     * @param LibraryRelease $releasePlanNode Node in release plan being released
+     */
+    protected function mergeLibrary(OutputInterface $output, LibraryRelease $releasePlanNode)
+    {
+        $library = $releasePlanNode->getLibrary();
+        $releaseBranch = $library->getBranch();
+        $minorBranch = str_replace(RewriteReleaseBranches::RELEASE_SUFFIX, '', $releaseBranch);
```

Comment:
What would a release branch look like for a patch release? Would we even have one? If so, would it be e.g. 4.12-release for a 4.12.6 patch release? If so this logic works fine for that scenario, I think?

For a major release, would that not just effectively be the same as a minor as far as branches (e.g. create 5.0-release branch, merge up to 5.0 and that up to 5)?

Reply:
I honestly don't know, I've only every used cow end to end for minor releases

I suppose for patch releases it would be from the same 4.12 branch and we'd still need to update and revert the stable deps and then do the double merge-up

For majors ... I really don't know if it uses 4 or 4.12 as the base branch before going to 5 / 5.0

Reply:
Okay. So we can agree that for patch releases it's the same flow as for the minor release (e.g. releasing a 4.12.999 patch needs the same stuff to happen as releasing a 4.12.0 minor as far as this PR is concerned).

I think that's true for majors as well. Releasing 5.0.0 is the same as releasing 5.1.0.
- a new `5.0` branch is created from `5`
- a new `5.0-release` branch is created from `5.0`
- constraints are updated in the `5.0-release` branch to be `~5.0.0@stable` (after having been `@beta` and `@rc`)
- all the normal release stuff happens. If there are any changes needed in this part, it's beyond the concerns of this PR
- after the release, merge-up starts
- merge up from `5.0-release` to `5.0` and destabalise constraints to `5.0.x-dev`
- merge up from `5.0` to `5` and destabalise constraints to `5.x-dev`
- delete release branch

I don't think this PR needs to change _at all_ for all of that to work.

Reply:
Right, sorry I think I confused myself with "releasing majors".  In hindsight it's very obviously going to use the existing `5` branch, just as it uses the existing `4` branch for minors.  We've created these branches ourselves.  There's possibly some existing functionality within cow where it would create the new major branches, but we won't use this anytime soon.